File: tests/test_wire_evolution_interop.py
#!/usr/bin/env python3
"""
Cross-version *extension interop* tests.

Unlike ``test_wire_evolution.py`` (which exercises a single schema and *simulates*
an older peer by trimming bytes), this suite generates two **separate** schema
versions of the same messages into two **separate** namespaces:

  * ``wire_evolution_v1`` -- base fields only, completely unaware of extensions.
  * ``wire_evolution_v2`` -- same msg_ids/base fields **plus** ``extensions_start``
    and the extension fields/variants.

Because each side is generated independently, neither knows about the other's
extra fields -- exactly mirroring two real code-bases that evolved apart.  The
two builds nevertheless agree on the magic bytes (computed from base fields
only) and the base field layout, which is what makes cross-version interop
work over a length-bearing profile.

Scenarios (mirrors the project's wire-evolution interop plan):

  1. Newer sender -> older receiver: v2 encodes extensions, v1 decodes base,
     trailing extension bytes skipped via the frame length field.
  2. Older sender -> newer receiver: v1 encodes base-only, v2 decodes base and
     zero-fills the extension fields.
  3. Same-version sanity: v2 -> v2 with extensions round-trips fully.
  4. Newer ext oneof variant -> older receiver degrades gracefully (unknown
     discriminator, no corruption).
  5. Older base oneof variant -> newer receiver decodes correctly.
  6. Multi-oneof: base oneof unaffected while the ext oneof exercises 4 and 5.
  7. Corrupted extension bytes invalidate the full CRC (magic seed still matches).
  8. Truncated payload (fewer bytes than the length field claims) is rejected.
  9. Length-less profile guard (IPC/Sensor): extension fields are fixed bytes,
     so a base-only older frame is *not* silently mis-parsed by a newer
     receiver -- documenting the "both sides must agree" constraint.
 10. Variable base array + trailing extension field, both interop directions.
"""
from __future__ import annotations

import math
import sys
from pathlib import Path
from test_utils import _check, run_generator, load_generated_module, REPO_ROOT, SRC_DIR

PROTO_V1 = REPO_ROOT / "tests" / "proto" / "wire_evolution_v1.sf"
PROTO_V2 = REPO_ROOT / "tests" / "proto" / "wire_evolution_v2.sf"

# Length-bearing profiles (carry an explicit payload length field).
LENGTH_BEARING = ("standard", "bulk", "network")
# Length-less profiles (fixed size; both sides must agree on the full size).
LENGTH_LESS = ("sensor", "ipc")


def _generate(out_dir: Path) -> None:
    """Generate the v1 and v2 schemas into the *same* output dir."""
    for proto in (PROTO_V1, PROTO_V2):
        result = run_generator(
            proto,
            "--build_py", "--py_path", str(out_dir / "py"), "--force",
        )
        if result.returncode != 0:
            print(result.stdout)
            print(result.stderr, file=sys.stderr)
            _check(False, f"code generation failed for {proto.name}")


def _load_gen(py_dir: Path, module_name: str):
    return load_generated_module(
        py_dir / "struct_frame" / "generated" / f"{module_name}.py",
        module_name,
    )


# ---------------------------------------------------------------------------
# Framing helpers (real length-bearing / length-less profile encode + parse)
# ---------------------------------------------------------------------------

class Framing:
    """Thin wrapper over the generated frame_profiles module."""

    def __init__(self, fp):
        self._configs = {
            "standard": fp.PROFILE_STANDARD_CONFIG,
            "sensor":   fp.PROFILE_SENSOR_CONFIG,
            "ipc":      fp.PROFILE_IPC_CONFIG,
            "bulk":     fp.PROFILE_BULK_CONFIG,
            "network":  fp.PROFILE_NETWORK_CONFIG,
        }
        self._fp = fp

    def encode(self, profile: str, msg) -> bytes:
        return self._fp.encode_message(self._configs[profile], msg)

    def parse(self, profile: str, buf: bytes, get_message_info):
        config = self._configs[profile]
        reader = self._fp.AccumulatingReader(config, get_message_info=get_message_info)
        reader.add_data(buf)
        return reader.next()

    def header_size(self, profile: str) -> int:
        return self._configs[profile].header_size


def _decode_zerofill(cls, info):
    """Decode a frame payload, zero-filling any extension bytes the sender omitted."""
    data = info.msg_data
    if len(data) < cls.MAX_SIZE:
        data = data + b"\x00" * (cls.MAX_SIZE - len(data))
    return cls.deserialize(data)


# ---------------------------------------------------------------------------
# Scenarios
# ---------------------------------------------------------------------------

def scenario_1_newer_to_older(v1, v2, fr: Framing) -> None:
    """Newer sender -> older receiver over every length-bearing profile."""
    for profile in LENGTH_BEARING:
        orig = v2.BaseExtensionMessage(header=0xBEEF, seq=42, crc_seed=0xDEADC0DE)
        buf = fr.encode(profile, orig)
        info = fr.parse(profile, buf, v1.get_message_info)
        _check(info.valid, f"[S1/{profile}] v2->v1 frame validates CRC (magic from base only)")
        _check(info.msg_len == v2.BaseExtensionMessage.MAX_SIZE,
               f"[S1/{profile}] frame length carries full v2 payload size")
        decoded = v1.BaseExtensionMessage.deserialize(info.msg_data)
        _check(decoded.header == 0xBEEF and decoded.seq == 42,
               f"[S1/{profile}] v1 decodes base fields; trailing extension bytes ignored")


def scenario_2_older_to_newer(v1, v2, fr: Framing) -> None:
    """Older sender -> newer receiver: extensions zero-filled."""
    orig = v1.BaseExtensionMessage(header=0x1234, seq=7)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v2.get_message_info)
    _check(info.valid, "[S2] v1->v2 base-only frame validates CRC")
    _check(info.msg_len == v1.BaseExtensionMessage.MAX_SIZE,
           "[S2] frame length carries shorter base-only payload")
    decoded = _decode_zerofill(v2.BaseExtensionMessage, info)
    _check(decoded.header == 0x1234 and decoded.seq == 7,
           "[S2] v2 decodes base fields correctly")
    _check(decoded.crc_seed == 0, "[S2] v2 extension field zero-filled to default")


def scenario_3_same_version(v2, fr: Framing) -> None:
    """Same-version v2 -> v2 with extensions populated round-trips fully."""
    cmd = v2.ExtCommandC(value_c=3.14, mode_c=2)
    orig = v2.OneOfExtensionMessage(device_id=2, command={"cmd_c": cmd},
                                    command_which="cmd_c", command_discriminator=3)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v2.get_message_info)
    _check(info.valid, "[S3] v2->v2 (ext variant) frame validates CRC")
    decoded = v2.OneOfExtensionMessage.deserialize(info.msg_data)
    _check(decoded.command_which == "cmd_c", "[S3] v2 round-trips extension variant discriminator")
    _check(math.isclose(decoded.command["cmd_c"].value_c, 3.14, rel_tol=1e-5),
           "[S3] v2 round-trips extension variant payload")


def scenario_4_newer_ext_variant_to_older(v1, v2, fr: Framing) -> None:
    """Newer ext oneof variant -> older receiver degrades gracefully."""
    cmd = v2.ExtCommandD(value_d=2.718281828)
    orig = v2.OneOfExtensionMessage(device_id=9, command={"cmd_d": cmd},
                                    command_which="cmd_d", command_discriminator=4)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v1.get_message_info)
    _check(info.valid, "[S4] v2 ext-variant -> v1 frame validates CRC")
    decoded = v1.OneOfExtensionMessage.deserialize(info.msg_data)
    _check(decoded.device_id == 9, "[S4] v1 still decodes the base header field")
    _check(decoded.command_which is None,
           "[S4] v1 treats unknown ext discriminator as no active variant (graceful)")
    _check(decoded.command_discriminator == 4,
           "[S4] v1 preserves the raw discriminator value without corruption")


def scenario_5_older_base_variant_to_newer(v1, v2, fr: Framing) -> None:
    """Older base oneof variant -> newer receiver decodes correctly."""
    cmd = v1.BaseCommandB(value_b=-1234, active_b=True)
    orig = v1.OneOfExtensionMessage(device_id=5, command={"cmd_b": cmd},
                                    command_which="cmd_b", command_discriminator=2)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v2.get_message_info)
    _check(info.valid, "[S5] v1 base-variant -> v2 frame validates CRC")
    decoded = _decode_zerofill(v2.OneOfExtensionMessage, info)
    _check(decoded.command_which == "cmd_b" and decoded.command["cmd_b"].value_b == -1234,
           "[S5] v2 decodes the base oneof variant correctly")


def scenario_6_multi_oneof(v1, v2, fr: Framing) -> None:
    """Multi-oneof: base oneof unaffected; ext oneof exercises S4/S5."""
    # Newer sender selects an extension variant in the *second* oneof only.
    first = v2.BaseCommandA(value_a=100, flags_a=5)
    sec = v2.ExtCommandC(value_c=2.71, mode_c=1)
    orig = v2.MultiOneOfExtensionMessage(
        priority=3,
        base_union={"first_a": first}, base_union_which="first_a", base_union_discriminator=1,
        ext_union={"second_ext": sec}, ext_union_which="second_ext", ext_union_discriminator=2,
    )
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v1.get_message_info)
    _check(info.valid, "[S6] v2 multi-oneof (ext in 2nd union) -> v1 validates CRC")
    decoded = v1.MultiOneOfExtensionMessage.deserialize(info.msg_data)
    _check(decoded.priority == 3, "[S6] v1 decodes base header field")
    _check(decoded.base_union_which == "first_a"
           and decoded.base_union["first_a"].value_a == 100,
           "[S6] v1 decodes the FIRST (base) oneof unaffected by ext in the second")
    _check(decoded.ext_union_which is None,
           "[S6] v1 treats the second oneof's unknown ext variant as inactive (graceful)")

    # Older sender selects the base variant in the second oneof -> newer receiver.
    sec_base = v1.BaseCommandA(value_a=77, flags_a=9)
    orig2 = v1.MultiOneOfExtensionMessage(
        priority=4,
        base_union={"first_b": v1.BaseCommandB(value_b=3, active_b=False)},
        base_union_which="first_b", base_union_discriminator=2,
        ext_union={"second_a": sec_base}, ext_union_which="second_a", ext_union_discriminator=1,
    )
    buf2 = fr.encode("standard", orig2)
    info2 = fr.parse("standard", buf2, v2.get_message_info)
    _check(info2.valid, "[S6] v1 multi-oneof (base in 2nd union) -> v2 validates CRC")
    decoded2 = _decode_zerofill(v2.MultiOneOfExtensionMessage, info2)
    _check(decoded2.ext_union_which == "second_a"
           and decoded2.ext_union["second_a"].value_a == 77,
           "[S6] v2 decodes the older base variant in the ext oneof correctly")


def scenario_7_corrupted_extension(v1, v2, fr: Framing) -> None:
    """Corrupted extension bytes invalidate the full CRC even though the
    magic (base seed) still matches."""
    orig = v2.BaseExtensionMessage(header=0xBEEF, seq=42, crc_seed=0xDEADC0DE)
    buf = bytearray(fr.encode("standard", orig))
    ext_offset = fr.header_size("standard") + v2.BaseExtensionMessage.BASE_SIZE
    buf[ext_offset] ^= 0xFF  # flip a byte inside the extension region
    info = fr.parse("standard", bytes(buf), v1.get_message_info)
    _check(not info.valid, "[S7] corrupted extension byte invalidates full CRC")


def scenario_8_truncated(v1, fr: Framing) -> None:
    """A payload shorter than its declared length field is rejected."""
    orig = v1.BaseExtensionMessage(header=1, seq=2)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf[:-1], v1.get_message_info)
    _check(not info.valid, "[S8] truncated payload (fewer bytes than length claims) is rejected")


def scenario_9_lengthless_guard(v1, v2, fr: Framing) -> None:
    """Length-less profiles: extensions are fixed bytes, both sides must agree."""
    # Positive control: same fixed size on both ends still works.
    same = v2.BaseExtensionMessage(header=0xAA, seq=5, crc_seed=0x99)
    info = fr.parse("ipc", fr.encode("ipc", same), v2.get_message_info)
    _check(info.valid, "[S9] IPC (length-less) same-version frame validates")

    # Cross-version base-only frame on a length-less profile: must be rejected.
    older = v1.BaseExtensionMessage(header=0xAA, seq=5)
    info = fr.parse("ipc", fr.encode("ipc", older), v2.get_message_info)
    _check(not info.valid,
           "[S9] IPC base-only older frame is rejected by newer receiver "
           "(both sides must agree on size)")


def scenario_10_variable(v1, v2, fr: Framing) -> None:
    """Variable base array + trailing extension field, both directions."""
    # Newer -> older: v1 reads the variable base, ignores the trailing ext bytes.
    orig = v2.VariableExtensionMessage(node_id=7, readings=[10, 20, 30],
                                       ext_timestamp=0x12345678)
    buf = fr.encode("standard", orig)
    info = fr.parse("standard", buf, v1.get_message_info)
    _check(info.valid, "[S10] v2 variable+ext -> v1 validates CRC")
    decoded = v1.VariableExtensionMessage.deserialize(info.msg_data)
    _check(decoded.node_id == 7 and list(decoded.readings) == [10, 20, 30],
           "[S10] v1 locates/decodes variable base; trailing ext bytes ignored")

    # Older -> newer: v2 zero-fills the extension after the variable base.
    orig2 = v1.VariableExtensionMessage(node_id=9, readings=[1, 2])
    buf2 = fr.encode("standard", orig2)
    info2 = fr.parse("standard", buf2, v2.get_message_info)
    _check(info2.valid, "[S10] v1 variable (base-only) -> v2 validates CRC")
    decoded2 = _decode_zerofill(v2.VariableExtensionMessage, info2)
    _check(decoded2.node_id == 9 and list(decoded2.readings) == [1, 2],
           "[S10] v2 locates variable base after cross-version decode")
    _check(decoded2.ext_timestamp == 0,
           "[S10] v2 zero-fills the trailing extension field")


# ---------------------------------------------------------------------------
# Single collected pytest entry point
# ---------------------------------------------------------------------------

def test_wire_evolution_interop(tmp_path: Path) -> None:
    """Cross-version wire-evolution interop: 10 scenarios verifying framing interop."""
    out = tmp_path
    _generate(out)
    py_dir = out / "py"

    fp = load_generated_module(py_dir / "frame_profiles.py", "frame_profiles")
    fr = Framing(fp)
    v1 = _load_gen(py_dir, "wire_evolution_v1")
    v2 = _load_gen(py_dir, "wire_evolution_v2")

    # Sanity: the two independent builds agree on magic + base size.
    for name in ("BaseExtensionMessage", "OneOfExtensionMessage",
                 "MultiOneOfExtensionMessage", "VariableExtensionMessage"):
        c1, c2 = getattr(v1, name), getattr(v2, name)
        _check((c1.MAGIC1, c1.MAGIC2) == (c2.MAGIC1, c2.MAGIC2),
               f"[setup] {name}: v1/v2 magic bytes agree (base-only seed)")
        _check(c1.BASE_SIZE == c2.BASE_SIZE,
               f"[setup] {name}: v1/v2 base_size agrees")

    scenario_1_newer_to_older(v1, v2, fr)
    scenario_2_older_to_newer(v1, v2, fr)
    scenario_3_same_version(v2, fr)
    scenario_4_newer_ext_variant_to_older(v1, v2, fr)
    scenario_5_older_base_variant_to_newer(v1, v2, fr)
    scenario_6_multi_oneof(v1, v2, fr)
    scenario_7_corrupted_extension(v1, v2, fr)
    scenario_8_truncated(v1, fr)
    scenario_9_lengthless_guard(v1, v2, fr)
    scenario_10_variable(v1, v2, fr)

File: tests/c/test_wire_evolution_interop.c
/**
 * Cross-version wire-evolution *interop* tests for the generated C bindings.
 *
 * Unlike a single-schema round-trip test, this runner uses two SEPARATELY
 * generated schema versions of the same messages:
 *
 *   - wire_evolution_v1  -- base fields only (older, extension-unaware)
 *   - wire_evolution_v2  -- base fields + `extensions_start` extension fields
 *
 * Each side has its own message structs, magic constants and
 * `*_get_message_info` lookup, exactly as two independently-built code-bases
 * would.  The tests exercise genuine newer<->older interop over the
 * length-bearing Standard/Bulk/Network profiles plus negative coverage.
 *
 * Scenarios (mirrors the project's wire-evolution interop plan, 1-7):
 *   1. Newer sender -> older receiver (v2 encodes extensions, v1 decodes base)
 *   2. Older sender -> newer receiver (v1 base-only, v2 zero-fills extensions)
 *   3. Same-version sanity (v2 -> v2 with extension variant)
 *   4. Newer ext oneof variant -> older receiver degrades gracefully
 *   5. Older base oneof variant -> newer receiver decodes correctly
 *   6. Multi-oneof: base oneof unaffected while ext oneof exercises 4/5
 *   7. Corrupted extension bytes invalidate the full CRC
 *
 * Build:
 *   gcc -I../../generated/c test_wire_evolution_interop.c -o test_wire_evolution_interop -lm
 */

#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdint.h>
#include <stdbool.h>

#include "wire_evolution_v1.structframe.h"
#include "wire_evolution_v2.structframe.h"
#include "frame_profiles.h"

static int g_pass = 0;
static int g_fail = 0;

static void check(bool cond, const char* msg) {
    if (cond) { printf("  [PASS] %s\n", msg); ++g_pass; }
    else      { printf("  [FAIL] %s\n", msg); ++g_fail; }
}

/* Encode a payload into a frame with extension-aware CRC over a profile. */
static size_t encode_ext(const profile_config_t* config, uint8_t* buffer, size_t cap,
                         uint8_t msg_id, const uint8_t* payload, size_t payload_size,
                         size_t base_size, uint8_t magic1, uint8_t magic2) {
    buffer_writer_t writer;
    buffer_writer_init(&writer, config, buffer, cap);
    return buffer_writer_write_ext(&writer, msg_id, payload, payload_size,
                                   0, 0, 0, 0, base_size, magic1, magic2);
}

static frame_msg_info_t parse_with(const profile_config_t* config, uint8_t* buffer,
                                   size_t frame_size, bool (*gmi)(uint16_t, message_info_t*)) {
    buffer_reader_t reader;
    buffer_reader_init(&reader, config, buffer, frame_size, gmi);
    return buffer_reader_next(&reader);
}

/* -------------------------------------------------------------------------
 * Scenario 1: newer sender -> older receiver over length-bearing profiles
 * ------------------------------------------------------------------------- */
static void scenario_1(void) {
    const profile_config_t* configs[3] = {
        &PROFILE_STANDARD_CONFIG, &PROFILE_BULK_CONFIG, &PROFILE_NETWORK_CONFIG };
    const char* names[3] = { "standard", "bulk", "network" };

    for (int i = 0; i < 3; ++i) {
        WireEvolutionV2BaseExtensionMessage m = {0};
        m.header = 0xBEEF; m.seq = 42; m.crc_seed = 0xDEADC0DE;
        uint8_t payload[64];
        size_t plen = WireEvolutionV2BaseExtensionMessage_serialize(&m, payload);

        uint8_t buffer[128];
        size_t fs = encode_ext(configs[i], buffer, sizeof(buffer),
                               WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MSG_ID,
                               payload, plen,
                               WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_BASE_SIZE,
                               WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MAGIC1,
                               WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MAGIC2);
        char label[96];
        snprintf(label, sizeof(label), "[S1/%s] v2->v1 frame validates CRC (magic from base)", names[i]);

        frame_msg_info_t r = parse_with(configs[i], buffer, fs, wire_evolution_v1_get_message_info);
        check(r.valid, label);

        if (r.valid && r.msg_data) {
            WireEvolutionV1BaseExtensionMessage d = {0};
            WireEvolutionV1BaseExtensionMessage_deserialize(r.msg_data, r.msg_len, &d);
            snprintf(label, sizeof(label), "[S1/%s] v1 decodes base; trailing ext bytes ignored", names[i]);
            check(d.header == 0xBEEF && d.seq == 42, label);
        }
    }
}

/* -------------------------------------------------------------------------
 * Scenario 2: older sender -> newer receiver (zero-fill extensions)
 * ------------------------------------------------------------------------- */
static void scenario_2(void) {
    WireEvolutionV1BaseExtensionMessage m = {0};
    m.header = 0x1234; m.seq = 7;
    uint8_t payload[64];
    size_t plen = WireEvolutionV1BaseExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V1_BASE_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V1_BASE_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V1_BASE_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V1_BASE_EXTENSION_MESSAGE_MAGIC2);

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v2_get_message_info);
    check(r.valid, "[S2] v1->v2 base-only frame validates CRC");

    if (r.valid && r.msg_data) {
        /* Newer receiver zero-fills the missing extension bytes before decoding. */
        uint8_t padded[WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MAX_SIZE] = {0};
        size_t copy_len = r.msg_len < sizeof(padded) ? r.msg_len : sizeof(padded);
        memcpy(padded, r.msg_data, copy_len);
        WireEvolutionV2BaseExtensionMessage d = {0};
        WireEvolutionV2BaseExtensionMessage_deserialize(padded, sizeof(padded), &d);
        check(d.header == 0x1234 && d.seq == 7, "[S2] v2 decodes base fields correctly");
        check(d.crc_seed == 0, "[S2] v2 extension field zero-filled to default");
    }
}

/* -------------------------------------------------------------------------
 * Scenario 3: same-version sanity (v2 -> v2 with extension variant)
 * ------------------------------------------------------------------------- */
static void scenario_3(void) {
    WireEvolutionV2OneOfExtensionMessage m = {0};
    m.device_id = 2;
    m.command_discriminator = WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_COMMAND_FIELD_CMD_C;
    m.command.cmd_c.value_c = 3.14f;
    m.command.cmd_c.mode_c = 2;
    uint8_t payload[64];
    size_t plen = WireEvolutionV2OneOfExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MAGIC2);

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v2_get_message_info);
    check(r.valid, "[S3] v2->v2 (ext variant) frame validates CRC");
    if (r.valid && r.msg_data) {
        WireEvolutionV2OneOfExtensionMessage d = {0};
        WireEvolutionV2OneOfExtensionMessage_deserialize(r.msg_data, r.msg_len, &d);
        check(d.command_discriminator == WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_COMMAND_FIELD_CMD_C
              && fabsf(d.command.cmd_c.value_c - 3.14f) < 1e-4f,
              "[S3] v2 round-trips the extension variant");
    }
}

/* -------------------------------------------------------------------------
 * Scenario 4: newer ext oneof variant -> older receiver degrades gracefully
 * ------------------------------------------------------------------------- */
static void scenario_4(void) {
    WireEvolutionV2OneOfExtensionMessage m = {0};
    m.device_id = 9;
    m.command_discriminator = WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_COMMAND_FIELD_CMD_D;
    m.command.cmd_d.value_d = 2.718281828;
    uint8_t payload[64];
    size_t plen = WireEvolutionV2OneOfExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MAGIC2);

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v1_get_message_info);
    check(r.valid, "[S4] v2 ext-variant -> v1 frame validates CRC");
    if (r.valid && r.msg_data) {
        WireEvolutionV1OneOfExtensionMessage d = {0};
        WireEvolutionV1OneOfExtensionMessage_deserialize(r.msg_data, r.msg_len, &d);
        check(d.device_id == 9, "[S4] v1 still decodes the base header field");
        /* v1 knows nothing about discriminator 4; it must preserve the raw value
         * without corrupting the base field. */
        check((int)d.command_discriminator == 4,
              "[S4] v1 preserves unknown ext discriminator without corruption");
    }
}

/* -------------------------------------------------------------------------
 * Scenario 5: older base oneof variant -> newer receiver decodes correctly
 * ------------------------------------------------------------------------- */
static void scenario_5(void) {
    WireEvolutionV1OneOfExtensionMessage m = {0};
    m.device_id = 5;
    m.command_discriminator = WIRE_EVOLUTION_V1_ONE_OF_EXTENSION_MESSAGE_COMMAND_FIELD_CMD_B;
    m.command.cmd_b.value_b = -1234;
    m.command.cmd_b.active_b = true;
    uint8_t payload[64];
    size_t plen = WireEvolutionV1OneOfExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V1_ONE_OF_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V1_ONE_OF_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V1_ONE_OF_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V1_ONE_OF_EXTENSION_MESSAGE_MAGIC2);

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v2_get_message_info);
    check(r.valid, "[S5] v1 base-variant -> v2 frame validates CRC");
    if (r.valid && r.msg_data) {
        uint8_t padded[WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_MAX_SIZE] = {0};
        size_t copy_len = r.msg_len < sizeof(padded) ? r.msg_len : sizeof(padded);
        memcpy(padded, r.msg_data, copy_len);
        WireEvolutionV2OneOfExtensionMessage d = {0};
        WireEvolutionV2OneOfExtensionMessage_deserialize(padded, sizeof(padded), &d);
        check(d.command_discriminator == WIRE_EVOLUTION_V2_ONE_OF_EXTENSION_MESSAGE_COMMAND_FIELD_CMD_B
              && d.command.cmd_b.value_b == -1234,
              "[S5] v2 decodes the older base oneof variant correctly");
    }
}

/* -------------------------------------------------------------------------
 * Scenario 6: multi-oneof, ext only in the second union
 * ------------------------------------------------------------------------- */
static void scenario_6(void) {
    /* Newer sender selects the ext variant in the second union only. */
    WireEvolutionV2MultiOneOfExtensionMessage m = {0};
    m.priority = 3;
    m.base_union_discriminator = WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_BASE_UNION_FIELD_FIRST_A;
    m.base_union.first_a.value_a = 100;
    m.base_union.first_a.flags_a = 5;
    m.ext_union_discriminator = WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_EXT_UNION_FIELD_SECOND_EXT;
    m.ext_union.second_ext.value_c = 2.71f;
    m.ext_union.second_ext.mode_c = 1;
    uint8_t payload[64];
    size_t plen = WireEvolutionV2MultiOneOfExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V2_MULTI_ONE_OF_EXTENSION_MESSAGE_MAGIC2);

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v1_get_message_info);
    check(r.valid, "[S6] v2 multi-oneof (ext in 2nd union) -> v1 validates CRC");
    if (r.valid && r.msg_data) {
        WireEvolutionV1MultiOneOfExtensionMessage d = {0};
        WireEvolutionV1MultiOneOfExtensionMessage_deserialize(r.msg_data, r.msg_len, &d);
        check(d.priority == 3
              && d.base_union_discriminator == WIRE_EVOLUTION_V1_MULTI_ONE_OF_EXTENSION_MESSAGE_BASE_UNION_FIELD_FIRST_A
              && d.base_union.first_a.value_a == 100,
              "[S6] v1 decodes the FIRST (base) oneof unaffected by ext in the second");
    }
}

/* -------------------------------------------------------------------------
 * Scenario 7: corrupted extension bytes invalidate the full CRC
 * ------------------------------------------------------------------------- */
static void scenario_7(void) {
    WireEvolutionV2BaseExtensionMessage m = {0};
    m.header = 0xBEEF; m.seq = 42; m.crc_seed = 0xDEADC0DE;
    uint8_t payload[64];
    size_t plen = WireEvolutionV2BaseExtensionMessage_serialize(&m, payload);

    uint8_t buffer[128];
    size_t fs = encode_ext(&PROFILE_STANDARD_CONFIG, buffer, sizeof(buffer),
                           WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MSG_ID,
                           payload, plen,
                           WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_BASE_SIZE,
                           WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MAGIC1,
                           WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_MAGIC2);

    /* Footer is 2 bytes; the extension region starts BASE_SIZE bytes into the
     * payload, which itself starts (frame_size - footer - msg_len) bytes in. */
    frame_msg_info_t probe = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v1_get_message_info);
    size_t payload_off = fs - 2 - probe.msg_len;
    size_t ext_off = payload_off + WIRE_EVOLUTION_V2_BASE_EXTENSION_MESSAGE_BASE_SIZE;
    buffer[ext_off] ^= 0xFF;

    frame_msg_info_t r = parse_with(&PROFILE_STANDARD_CONFIG, buffer, fs, wire_evolution_v1_get_message_info);
    check(!r.valid, "[S7] corrupted extension byte invalidates full CRC");
}

int main(void) {
    printf("=== C Cross-Version Wire-Evolution Interop Tests ===\n\n");
    printf("Scenario 1: newer sender -> older receiver (length-bearing)\n");
    scenario_1();
    printf("\nScenario 2: older sender -> newer receiver (zero-fill)\n");
    scenario_2();
    printf("\nScenario 3: same-version sanity (v2 -> v2)\n");
    scenario_3();
    printf("\nScenario 4: newer ext oneof variant -> older receiver\n");
    scenario_4();
    printf("\nScenario 5: older base oneof variant -> newer receiver\n");
    scenario_5();
    printf("\nScenario 6: multi-oneof (ext only in 2nd union)\n");
    scenario_6();
    printf("\nScenario 7: corrupted extension bytes invalidate CRC\n");
    scenario_7();
    printf("\nResults: %d passed, %d failed\n", g_pass, g_fail);
    return g_fail == 0 ? 0 : 1;
}
